pkg/appctx/container.go

```diff
@@ -0,0 +1,102 @@
+package appctx
+
+import (
+	"context"
+	"fmt"
+	"sync"
+)
+
+type containerKeyType int
+
+var containerKey containerKeyType = 1
+
+type ErrNoApplicationContainerFound struct{}
+
+func (e ErrNoApplicationContainerFound) Error() string {
+	return "no application container found in context"
+}
+
+type ErrNoItemFound struct {
+	key interface{}
+}
+
+func (e ErrNoItemFound) Error() string {
+	return fmt.Sprintf("no item with key %s found", e.key)
+}
+
+type container struct {
+	sync.Mutex
+	items map[interface{}]interface{}
+}
+
+func WithContainer(ctx context.Context) context.Context {
+	return context.WithValue(ctx, containerKey, &container{
+		items: make(map[interface{}]interface{}),
+	})
+}
+
+func Set(ctx context.Context, key interface{}, value interface{}) error {
+	contI := ctx.Value(containerKey)
+
+	if contI == nil {
+		return &ErrNoApplicationContainerFound{}
+	}
+
+	cont := contI.(*container)
+
+	cont.Lock()
+	defer cont.Unlock()
+
+	cont.items[key] = value
+
+	return nil
+}
+
+func Get(ctx context.Context, key interface{}) (interface{}, error) {
+	contI := ctx.Value(containerKey)
+
+	if contI == nil {
+		return nil, &ErrNoApplicationContainerFound{}
+	}
+
+	cont := contI.(*container)
+
+	cont.Lock()
+	defer cont.Unlock()
+
+	if val, ok := cont.items[key]; ok {
+		return val, nil
+	}
+
+	return nil, &ErrNoItemFound{
+		key: key,
+	}
+}
+
+func GetSet(ctx context.Context, key interface{}, factory func() (interface{}, error)) (interface{}, error) {
+	var ok bool
+	var err error
+	var contI, val interface{}
+
+	if contI = ctx.Value(containerKey); contI == nil {
+		return factory()
+		// return nil, &ErrNoApplicationContainerFound{}
```

Comment:
I don't think we should just return the factory result here, as you would expect when using `appctx` it actually uses the app context (making the entity a singleton, vs. no a singleton here)
Maybe you could just initialize an app ctx here?

I think it will be difficult to reason about if the app ctx is only partially available, I'd rather have it everywhere available

Reply:
Yes, this is a temporary solution, until the ctx handling gets cleaned up in a following PR. I don't want to add more changes to this one. Right now, applications will always have the appctx container, but eg our integration tests are a little bit more tricky in regards to this.